### ISSUE TITLE
5.9.5 resolves to GLOB constant, not subpackage name

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#8963]
Migrated from [rt.perl.org#43663](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=43663) (status was 'resolved')

Searchable as RT43663$

**Comment by p5pRT** [comment]
### From chmrr@zoq-fot-pik.scd

#### Created by chmrr@zoq-fot-pik.scd



The following test fails under 5\.9\.5\, but works under 5\.8\.8\.
UNIVERSAL&#8203;::require sees a GLOB as its first argument \(as a result of
comstant\.pm's meddlings\)\, instead of the class name\, causing it to
fail\.  Using 5\.8\.8's constant\.pm under perl 5\.9\.5 does work\.

\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\< \./test\.pl
package Foo;

use constant Bar => 'Something bad';

package main;
require UNIVERSAL&#8203;::require;

my $class = "Foo&#8203;::Bar";
eval \{ $class\->require \} or print "not ";
print "ok\\n";
\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\< end \./test\.pl

\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\< \./Foo/Bar\.pm
package Foo&#8203;::Bar;
1;
\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\< end \./Foo/Bar\.pm

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=medium

Site configuration information for perl 5.9.5:

Configured by chmrr at Sun Jul  8 16:56:59 EDT 2007.

Summary of my perl5 (revision 5 version 9 subversion 5) configuration:
  Platform:
    osname=linux, osvers=2.6.21-suspend2-r4, archname=i686-linux
    uname='linux zoq-fot-pik 2.6.21-suspend2-r4 #2 preempt mon may 21 15:50:36 edt 2007 i686 intel(r) pentium(r) m processor 1400mhz genuineintel gnulinux '
    config_args='-DEBUGGING -Dusedevel -Doptimize=-ggdb3'
    hint=previous, useposix=true, d_sigaction=define
    useithreads=undef, usemultiplicity=undef
    useperlio=define, d_sfio=undef, uselargefiles=define, usesocks=undef
    use64bitint=undef, use64bitall=undef, uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cc', ccflags ='-fno-strict-aliasing -pipe -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-ggdb3',
    cppflags='-fno-strict-aliasing -pipe -I/usr/local/include -fno-strict-aliasing -pipe -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -fno-strict-aliasing -pipe -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64'
    ccversion='', gccversion='4.1.2 (Gentoo 4.1.2)', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=4, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib
    libs=-lnsl -lndbm -lgdbm -ldb -ldl -lm -lcrypt -lutil -lc
    perllibs=-lnsl -ldl -lm -lcrypt -lutil -lc
    libc=/lib/libc-2.5.so, so=so, useshrplib=false, libperl=libperl.a
    gnulibc_version='2.5'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-Wl,-E'
    cccdlflags='-fPIC', lddlflags='-shared -L/usr/local/lib'

Locally applied patches:
    


@INC for perl 5.9.5:
    /home/chmrr/work/bps/jifty/lib
    /home/chmrr/work/bps/Jifty-DBI/lib
    /home/chmrr/prog/perl/blead-install/lib/5.9.5/i686-linux
    /home/chmrr/prog/perl/blead-install/lib/5.9.5
    /home/chmrr/prog/perl/blead-install/lib/site_perl/5.9.5/i686-linux
    /home/chmrr/prog/perl/blead-install/lib/site_perl/5.9.5
    /home/chmrr/prog/perl/blead-install/lib/vendor_perl/5.9.5/i686-linux
    /home/chmrr/prog/perl/blead-install/lib/vendor_perl/5.9.5
    /home/chmrr/prog/perl/blead-install/lib/vendor_perl
    .


Environment for perl 5.9.5:
    HOME=/home/chmrr
    LANG=en_US.UTF-8
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/usr/local/bin:/usr/bin:/bin:/opt/bin:/usr/athena/bin:/usr/i686-pc-linux-gnu/gcc-bin/4.1.2:/usr/i686-pc-linux-gnu/gcc-bin/4.1.1:/opt/sun-jdk-1.4.2.14/bin:/opt/sun-jdk-1.4.2.14/jre/bin:/opt/sun-jdk-1.4.2.14/jre/javaws:/usr/games/bin:/home/chmrr/bin:/home/chmrr/bin
    PERL5LIB=/home/chmrr/work/bps/jifty/lib:/home/chmrr/work/bps/Jifty-DBI/lib
    PERL_BADLANG (unset)
    SHELL=/bin/bash

```
</details>

**Comment by p5pRT** [comment]
### From @cpansprout

This was fixed by&#8203;:

From 5e0caaeb4056614af54deb646e72147bdc6dcadc Mon Sep 17 00&#8203;:00&#8203;:00 2001
From&#8203;: Nicholas Clark \<nick@&#8203;ccl4\.org>
Date&#8203;: Mon\, 22 Oct 2007 20&#8203;:19&#8203;:55 \+0000
Subject&#8203;: \[PATCH\] SQL&#8203;::Translator triggered a subtle piece of wrongness\,
whereby it
 managed to cause a proxy constant subroutine in the symbol table to
 be upgraded to a GV at precisely the point where it was being
 considered as a possible IO handle name\. The core's user\-space visible
 behaviour certainly was buggy \- I believe this patch to be the correct
 fix\. However\, it seems to be very hard to write a test for it\.

p4raw\-id&#8203;: //depot/perl@&#8203;32172

Was a test ever added? We could base one on this&#8203;:

perl5\.9\.5 \-Mconstant=Just\,0 \-le 'sub
IO&#8203;::Handle&#8203;::a\_japh\{print\*\{$\_\[0\]\}=~/\(\[^&#8203;:\]\+\)$/\," another Perl hacker\,"\}
"Just"\->a\_japh'

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @cpansprout

On Sun Sep 26 13&#8203;:27&#8203;:13 2010\, sprout wrote&#8203;:
> This was fixed by&#8203;:
> 
> From 5e0caaeb4056614af54deb646e72147bdc6dcadc Mon Sep 17 00&#8203;:00&#8203;:00 2001
> From&#8203;: Nicholas Clark \<nick@&#8203;ccl4\.org>
> Date&#8203;: Mon\, 22 Oct 2007 20&#8203;:19&#8203;:55 \+0000
> Subject&#8203;: \[PATCH\] SQL&#8203;::Translator triggered a subtle piece of wrongness\,
> whereby it
>  managed to cause a proxy constant subroutine in the symbol table to
>  be upgraded to a GV at precisely the point where it was being
>  considered as a possible IO handle name\. The core's user\-space visible
>  behaviour certainly was buggy \- I believe this patch to be the correct
>  fix\. However\, it seems to be very hard to write a test for it\.
> 
> p4raw\-id&#8203;: //depot/perl@&#8203;32172
> 
> Was a test ever added? We could base one on this&#8203;:
> 
> perl5\.9\.5 \-Mconstant=Just\,0 \-le 'sub
> IO&#8203;::Handle&#8203;::a\_japh\{print\*\{$\_\[0\]\}=~/\(\[^&#8203;:\]\+\)$/\," another Perl hacker\,"\}
> "Just"\->a\_japh'

A test has been added in commit 7148157\.

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
@cpansprout - Status changed from 'open' to 'resolved'
